misk/src/main/kotlin/misk/cloud/gcp/storage/LocalStorageRpc.kt

```diff
@@ -0,0 +1,582 @@
+package misk.cloud.gcp.storage
+
+import com.google.api.services.storage.model.StorageObject
+import com.google.cloud.Tuple
+import com.google.cloud.storage.BlobId
+import com.google.cloud.storage.StorageException
+import com.google.cloud.storage.spi.v1.StorageRpc
+import com.google.cloud.storage.spi.v1.StorageRpc.Option.IF_GENERATION_MATCH
+import com.google.cloud.storage.spi.v1.StorageRpc.Option.IF_GENERATION_NOT_MATCH
+import com.squareup.moshi.Moshi
+import misk.io.listRecursively
+import misk.moshi.adapter
+import misk.nio.withLock
+import misk.okio.forEachBlock
+import okio.Okio
+import java.io.FileNotFoundException
+import java.io.IOException
+import java.io.InputStream
+import java.math.BigInteger
+import java.nio.ByteBuffer
+import java.nio.channels.FileChannel
+import java.nio.file.FileAlreadyExistsException
+import java.nio.file.Files
+import java.nio.file.Path
+import java.nio.file.Paths
+import java.nio.file.StandardCopyOption.ATOMIC_MOVE
+import java.nio.file.StandardCopyOption.REPLACE_EXISTING
+import java.nio.file.StandardOpenOption.CREATE
+import java.nio.file.StandardOpenOption.READ
+import java.nio.file.StandardOpenOption.TRUNCATE_EXISTING
+import java.nio.file.StandardOpenOption.WRITE
+import java.util.UUID
+import java.util.concurrent.locks.ReentrantReadWriteLock
+import kotlin.concurrent.read
+import kotlin.concurrent.write
+import kotlin.streams.asSequence
+
+/**
+ * Implementation of [StorageRpc] that is backed by local disk storage. Useful for running
+ * in development mode on local machines, since there is no GCS emulator available. Files
+ * are stored with the generation version appended as an extension suffix, with a symlink
+ * existing for the latest generation. To preserve GCS atomicity semantics, progressive
+ * uploads are handled by storing the interim uploaded data in a temporary file based
+ * off the latest generation, then doing a rename to the new generation once the upload
+ * is complete.
+ *
+ * The implementation uses file locks to support multiple local processes accessing the same
+ * storage directory. Whenever a blob is updated, the local store will acquire an exclusive
+ * lock on a corresponding lock file, releasing that lock when the update is complete (or when
+ * the process dies). It's slightly more complicated since we need to deal with the possibility
+ * that writer stops partway through without the process failing; in those cases we don't want to
+ * prevent subsequent updates from other processes.
```

Comment:
I wonder if this is too much, and just atomic renaming is the only concurrency primitive we should have.

(Atomic renaming is pretty limited)

Reply:
... oh and that approach is totally broken if you want to do generation numbers. Never mind me.

Reply:
Also uploads are broken into chunks, so you need to store the chunks in a temp location until all of them are received, at which point you can do the atomic move.